### PR TITLE
SAK-44623: No bullhorn notification is thrown when an assignment's Opening Date is in the future

### PR DESCRIPTION
**Pull request by JKJANETSCHEK**
implementation of bullhorn notifications when an assignment's Opening Date is in the future. Implementation is based on "asn.available.assignment" event.
Changes were also made to BaseEventDelayHandler.java because bullhorn notification execution of delayed event would build wrong DeepLink for users with no "asn.new" permission.

new pull request with updates.

### CODE REVIEW

File: assignment/api/src/java/org/sakaiproject/assignment/api/AssignmentConstants.java [hunk 1]
```diff
@@ -45,6 +45,12 @@ public final class AssignmentConstants {
      */
     public static final String EVENT_ADD_ASSIGNMENT = "asn.new.assignment";
 
+
+    /**
+     *  SAK-44623: event for delayed assignment
```
Comment (adrianfish) [inline assignment/api/src/java/org/sakaiproject/assignment/api/AssignmentConstants.java:50]: I wouldn't bother with the SAK. We can get that from the git history and they just end up polluting the code, imo. They seem a good idea, until everybody starts doing them :)

File: assignment/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java [hunk 2]
```diff
@@ -8658,10 +8670,28 @@ private void post_save_assignment(RunData data, String postOrSave) {
                     }
                 }
             }
-            //SAK-45967
-            if (newAssignment || !a.getDraft()) {
-                // post new assignment event since it is fully initialized by now
-                eventTrackingService.post(eventTrackingService.newEvent(AssignmentConstants.EVENT_ADD_ASSIGNMENT, assignmentReference, true));
+            // SAK-45967 && SAK-44623 && SAK-45992
```
Comment (adrianfish) [inline assignment/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java:8673]: I'd just get rid of this line. We can work this stuff out from the git history.

File: assignment/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java [hunk 3]
```diff
@@ -8658,10 +8670,28 @@ private void post_save_assignment(RunData data, String postOrSave) {
...
+            if ((newAssignment && !a.getDraft()) || (!a.getDraft() && !newAssignment)) {
+
+                Collection aGroups = a.getGroups();
+                if (aGroups.size() != 0) {
+                    //SAK-45992
```
Comment (adrianfish) [inline assignment/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java:8678]: I'd lose the SAK references. Maybe // If already open ?

File: assignment/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java [hunk 4]
```diff
@@ -8658,10 +8670,28 @@ private void post_save_assignment(RunData data, String postOrSave) {
...
+                    if (openTime.isBefore(Instant.now())) {
+                        eventTrackingService.post(eventTrackingService.newEvent(AssignmentConstants.EVENT_UPDATE_ASSIGNMENT_ACCESS, assignmentReference, true));
+                    } else {
+                        // SAK-44623
```
Comment (adrianfish) [inline assignment/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java:8682]: Maybe // Not open yet, delay the event ?

File: assignment/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java [hunk 5]
```diff
@@ -8658,10 +8670,28 @@ private void post_save_assignment(RunData data, String postOrSave) {
...
+                        eventTrackingService.delay(eventTrackingService.newEvent(AssignmentConstants.EVENT_AVAILABLE_ASSIGNMENT, assignmentReference,
+                                true), openTime);
+                    }
+                } else {
+                    //SAK-44623
+                    if (openTime.isBefore(Instant.now())) {
+                        // post new assignment event since it is fully initialized by now
+                        eventTrackingService.post(eventTrackingService.newEvent(AssignmentConstants.EVENT_ADD_ASSIGNMENT, assignmentReference, true));
```
Comment (adrianfish) [inline assignment/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java:8690]: Do you need to cancel any delayed events here?

File: content/content-impl-providers/impl/src/java/org/sakaiproject/content/providers/BaseEventDelayHandler.java [hunk 6]
```diff
@@ -308,12 +308,23 @@ public void execute(String opaqueContext)
...
 				securityService.pushAdvisor(new SecurityAdvisor() {
 				    public SecurityAdvice isAllowed(String userId, String function, String reference) {
-				            if (securityService.unlock(event.getUserId(), function, reference)) {
-				                return SecurityAdvice.ALLOWED;
-				             }
-				            return SecurityAdvice.PASS;
-				         }
-				    });
+						//SAK-44623: bullhorn notification execution of delayed "asn.available.assignment" event would build wrong DeepLink for users with no "asn.new" permission.
+						if(StringUtils.isNotEmpty(userId)){
```
Comment (adrianfish) [inline content/content-impl-providers/impl/src/java/org/sakaiproject/content/providers/BaseEventDelayHandler.java:312]: Indentation seems a bit off here.

**Review by adrianfish — CHANGES_REQUESTED** [review]
Just some small stuff.

**Review by adrianfish — CHANGES_REQUESTED** [review]

File: assignment/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java (hunk 2 above)
Comment (JKJANETSCHEK) [reply assignment/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java:8673]: done

File: assignment/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java (hunk 4 above)
Comment (JKJANETSCHEK) [reply assignment/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java:8682]: done

File: content/content-impl-providers/impl/pom.xml [hunk 7]
```diff
@@ -31,6 +31,11 @@
       <groupId>org.sakaiproject.scheduler</groupId>
       <artifactId>scheduler-api</artifactId>
     </dependency>
+    <dependency>
+      <groupId>org.apache.commons</groupId>
+      <artifactId>commons-lang3</artifactId>
+      <version>3.0</version>
```
Comment (ern) [inline content/content-impl-providers/impl/pom.xml:37]:
```suggestion
```
no need to import the version here as it is inherited from master

**Pull request merged**